virtcontainers/device/drivers/vfio.go

```diff
@@ -97,17 +100,61 @@ func (device *VFIODevice) DeviceType() config.DeviceType {
 	return device.DevType
 }
 
+type VFIODevice_Type int32
+
+const (
+        VFIODEVICE_TYPE_VFIO_ERROR    VFIODevice_Type = 0
+        VFIODEVICE_TYPE_VFIO_NORMAL   VFIODevice_Type = 1
+        VFIODEVICE_TYPE_VFIO_MEDIATED VFIODevice_Type = 2
+)
+
+func getVFIODetails(deviceFileName, iommuDevicesPath string) (deviceBDF, deviceSysfsDev string, vfioDeviceType VFIODevice_Type, err error) {
+
+        tokens := strings.Split(deviceFileName, ":")
+        vfioDeviceType = VFIODEVICE_TYPE_VFIO_ERROR
+        if len(tokens) == 3 {
+                vfioDeviceType = VFIODEVICE_TYPE_VFIO_NORMAL
+        } else {
+                tokens = strings.Split(deviceFileName, "-")
+                if len(tokens) == 5 {
+                        vfioDeviceType = VFIODEVICE_TYPE_VFIO_MEDIATED
+                }
+        }
+
+        err = nil
+        deviceBDF = ""
+        deviceSysfsDev = ""
+        if (vfioDeviceType == VFIODEVICE_TYPE_VFIO_NORMAL) {
+                //Get bdf of device eg 0000:00:1c.0
+                deviceBDF = getBDF(deviceFileName)
+        } else if (vfioDeviceType == VFIODEVICE_TYPE_VFIO_MEDIATED){
+                //Get sysfsdev of device eg f79944e4-5a3d-11e8-99ce-479cbab002e4
+                sysfsDevStr := filepath.Join(iommuDevicesPath, deviceFileName)
+                deviceSysfsDev, err = getSysfsDev(sysfsDevStr)
+        } else {
+                err = fmt.Errorf("Incorrect tokens found while parsing vfio details: %s", deviceFileName)
+        }
+
+        if err != nil {
+                return "", "", vfioDeviceType, err;
+        } else {
+                return deviceBDF, deviceSysfsDev, vfioDeviceType, nil;
+        }
+
+}
+
 // getBDF returns the BDF of pci device
 // Expected input strng format is [<domain>]:[<bus>][<slot>].[<func>] eg. 0000:02:10.0
-func getBDF(deviceSysStr string) (string, error) {
-	tokens := strings.Split(deviceSysStr, ":")
-
-	if len(tokens) != 3 {
-		return "", fmt.Errorf("Incorrect number of tokens found while parsing bdf for device : %s", deviceSysStr)
-	}
+func getBDF(deviceSysStr string) (string) {
+	tokens := strings.SplitN(deviceSysStr, ":", 2)
+	return tokens[1]
```

Comment:
please check length before returning

Reply:
The length has been checked when calculating vfioDeviceType in the code above.